x-pack/plugins/monitoring/server/kibana_monitoring/bulk_uploader.js

```diff
@@ -119,41 +121,36 @@ export class BulkUploader {
     return sendBulkPayload(this._client, this._interval, payload);
   }
 
-  static deepMergeUploadData(uploadData, collectorSet) {
-    const deepMergeAndGroup = collectorSet.bulkFormat(uploadData).reduce((accum, datas) => {
-      for (const data of datas) {
-        accum[data.type] = accum[data.type] || {};
-        for (const key in data.payload) {
-          if (typeof accum[data.type][key] === 'object') {
-            accum[data.type][key] = {
-              ...accum[data.type][key],
-              ...data.payload[key]
-            };
-          } else {
-            accum[data.type][key] = data.payload[key];
-          }
-        }
+  /*
+   * Bulk stats are transformed into a bulk upload format
+   * Non-legacy transformation is done in CollectorSet.toApiStats
+   */
+  toBulkUploadFormat(rawData, collectorSet) {
+    // convert the raw data to a nested object by taking each payload through
+    // its formatter, organizing it per-type
+    const typesNested = rawData.reduce((accum, { type, result }) => {
+      if (isEmpty(result)) {
+        return accum;
       }
-      return accum;
+      const { type: uploadType, payload: uploadData } = collectorSet.getCollectorByType(type).formatForBulkUpload(result);
+      return defaultsDeep(accum, { [uploadType]: uploadData });
     }, {});
 
-    return Object.keys(deepMergeAndGroup).reduce((accum, type) => {
-      accum.push([
+    // convert the nested object into a flat array, with each payload prefixed
+    // with an 'index' instruction, for bulk upload
+    const flat = Object.keys(typesNested).reduce((accum, type) => {
+      return [
+        ...accum,
         { index: { _type: type } },
-        deepMergeAndGroup[type],
-      ]);
-      return accum;
+        {
+          kibana: this._getKibanaInfoForStats(),
```

Comment:
I had to make `BulkUploader.toBulkUploadFormat` a non-static method to support `this`